test/misc.jl

```diff
@@ -573,6 +573,17 @@ end
 # issue #44780
 @test summarysize(BigInt(2)^1000) > summarysize(BigInt(2))
 
+# issue #53061
+mutable struct S53061
+    x::Union{Float64, Tuple{Float64, Float64}}
+    y::Union{Float64, Tuple{Float64, Float64}}
+end
+let s = S53061[S53061(rand(), (rand(),rand())) for _ in 1:10^5]
```

Comment:
not a big deal, but does the array have to be this big to reproduce this?

Reply:
checked, the answer is yes, at least in the current form of the struct, more you reduce the size more you need attempts to find them unequal

Reply:
e.g. this usually works at 10^4: `allequal([Base.summarysize(s) for i in 1:10]`.

Reply:
why is this nondeterministic? That makes very little sense to me.

Reply:
to me too actually, but I'm not qualified enough to answer :D

Reply:
It's because the old code has to allocate a box for the value, which it should not have to do, and so the later uniquing by address becomes dependent on GC/allocator behavior. The correct way to fix this is to iterate over the pointers inside a struct instead of over its fields (some fields might be inlined but contain embedded pointers).